### ISSUE TITLE
¿Why I don't have exposure checks since 2020-09-10 19:24:55?

### ISSUE THREAD

**Issue by CarlosJimenez** [RadarCOVID/radar-covid-ios#26]
### I installed the App at the end of Aug 2020.
The exposure Api was working fine. I checked in Setttings -> Privacy. and I was a happy user.
The app wasn't updated in App Store [from Aug 20 until Sep 16](https://apps.apple.com/es/app/radar-covid/id1520443509)
<img width="200" alt="updates" src="https://user-images.githubusercontent.com/1741754/93416674-83570800-f8a6-11ea-827c-e792c8f76ebd.png">

### Exposure Notification Express on iOS 13.7 

https://developer.apple.com/documentation/exposurenotification/supporting_exposure_notifications_express
A few days ago, iOS 13.7 update enhanced behaviour of the exposure Api. To support several country apps it moves the setting to a new place and allows to use more than one country specific app. So when you travel the exposure app automatically select the appropriate app based on location. Also brings a export option of the exposure checks.

![export](https://user-images.githubusercontent.com/1741754/93418444-b9968680-f8aa-11ea-9e6f-d7571ae0e995.jpg)

### The issue 
When I have seen today the exposure checks I realise that I didn't have notifications since a week ago even when I didn't change anything. I still had bluetooth enabled, I sill have the app installed. I didn't made any change on my side. Something has changed.
It could be on the RadarCovid servers or it could be that the update of Apple to support worldwide covid Apps had broken something. **Maybe the apps need to be updated to interact with the exposure notification express api** to see if they are in charge of notifications.

### My exposure checks JSON
I attach my exposure checks json [ExposureChecks-2020-09-17.json.zip](https://github.com/RadarCOVID/radar-covid-ios/files/5236105/ExposureChecks-2020-09-17.json.zip)
I kindly suggest to other members of GitHub to check (or export) their exposure checks json and verify if there is notification since the iOS 13.7 update. I would like that the issue only affects to me, but it could be possible that affects other users.

### Useful links
* https://covid19.apple.com/contacttracing
* https://developer.apple.com/documentation/exposurenotification/supporting_exposure_notifications_express

**Comment by jorgej-ramos** [comment]
App was updated yesterday to version 1.0.7. This update seems to fix several problems with iOS 13.7 and iOS 14. Here's part of my exposure checks
```
  {
      "Timestamp" : "2020-09-16 20:16:00 +0200",
      "Files" : [
        {
          "Hash" : "0BCF2[... HASH MASKED ...]9DE88",
          "MatchCount" : 0,
          "KeyCount" : 61,
          "AppBundleIdentifier" : "es.gob.radarcovid",
          "Timestamp" : "2020-09-16 20:16:00 +0200"
        }
      ],
      "AppBundleIdentifier" : "es.gob.radarcovid"
    }
```
Please, check your app version.

On the other hand, the Android colleagues maintain an open discussion about the two time windows that exist to download the data of new infections.

https://github.com/RadarCOVID/radar-covid-android/issues/1

**Comment by CarlosJimenez** [comment]
### As soon as I have a new exposure check I will write here to verify that the update fixed the issue
It's nice to know that the version 1.0.7 will fix the issue, 
I checked the version and updated the app a couple of hours ago (before open the ticket). I don't know why the app was not updated automatically even when I have automatic updated enabled. My last exposure check was at 1:15 am and I don't have a new one. As soon as the exposure check list will be updated I will write here. 
**Thank you for the link to the android colleagues, maybe I will have to wait until 1 pm.**
**UPDATE: At 8 pm I don't have any additional exposure check since 1:15 am.** I hope the latest version doesn't break everything.  

### But I think that the issue should be public
if the automatic updates is not working properly (maybe for the nature of the app) and the exposure checks are not working since the update of iOS 13.7 (a week ago) I think that the issue should be public to allow the users of the app to check if they have the update properly installed.

I also suggest a better communication with Apple and Google to prevent that future enhancements of the API will be supported by the app and not reacting late. This app could save lives (I lost my two parents in March) and I think that a week without sharing my exposure checks could compromise my health and the health of all people that I met last week

**Comment by borjam** [comment]
I think there is a serious issue because the application doesn't give a hint whether it's working properly or not.

I understand the designers want it to be as hassle free as possible, but discarding errors can make the application create a false sense of security, which will achieve exactly the opposite of its intended purpose.

Please review that behavior thoroughly. I would add some **documented behavior** so that, when installing it, the user can know that it's working. Something like a first data download or some test check.

**Comment by CarlosJimenez** [comment]
**FINAL UPDATE: At 10:49pm** I hadn't any additional exposure checks. So I decided uninstall and install again Radar Covid. After that, I got it a new exposition check. Not empty this time. So it's not enough to force an update of Radar Covid. You need to reinstall it. It's like a new enrollment enables the Api-App communication again.

The procedure is very complicated for most of the users. I am afraid that most of the millions of users of Radar Covid has an app that doesn't exchange exposure checks. Only new users with a clean setup of version 1.0.7 or users of previous versions that they didn't update iOS have a working exposure app. 

I guess that the crazy workflow of updating the app is outside of competence of RadarCovid developers. But RadarCovid team should inform to the millions of users that they should reinstall the app. If not, the app is useless.

**Comment by pvieito** [comment]
@CarlosJimenez There are two daily checks but the exact time is decided by the system and the hardware conditions. There is no need to reinstall the app. I have seen multiple examples of people using iOS 13.7 / iOS 14 (myself included) and once the app was updated to 1.0.7 it started working properly. In fact, on iOS 13.7 / iOS 14 it did indeed run the check twice a day, but the system would refuse the unsigned diagnosis keys, which was the bug fixed by DP3T SDK 1.2.1.

**Clarification**: *There is no need to reinstall the app* in the context of the [DP3T SDK 1.2.1 issue](https://github.com/RadarCOVID/radar-covid-ios/issues/4) or the [Radar COVID 1.0.6 issue](https://github.com/RadarCOVID/radar-covid-ios/issues/24). Of course, there may be other currently unknown bugs, which must be addressed.

**Comment by CarlosJimenez** [comment]
I assume that @pvieito is linked to the development team

> **@pvieito : There is no need to reinstall the app**

The app was installed by millions of users. I don't know how many multiple cases you have seen to think that there is no need to reinstall the app. I hope you don't have the syndrome "*If it works in my iPhone and 100 more, it will work in ALL iPhones*".
I am going to divide your statement in 2:

* **There is no need to manually force an update of the app Radar Covid**
In my case the app were not updated until I manually force the update. My partner the same case. Even we had automatic updates of the apps in App Store configuration. This behaviour is strange to me. 
Do you update automatically the app or your case is different from mine in this point?
Why the app were not automatically updated? 
  * If this affect to all users that updated iOS 13.7 without updating the app, it should be the main concern

* **Once the app Radar Covid is updated (automatically, manually?), it will start working properly**
I was expecting a new exposure check to report it here asap. I waited from 01:15 to 22:49 without success. I didn't wait one hour more to see if your statement "twice by day" were true. My fault.
I reboot the iPhone. No exposure checks.
I decided to reinstall Radar Covid. After reinstalling it I had a new exposure check immediately. 
I don't know if the app has control about exposure check triggering. If so, I suggest to force a exposure check when the app is updated. It will be useful in cases like mine after a week without valid exposure checks.
  * If you have to warn all users to update manually the app, It's less important if you ask for a manual update or reinstalling the app

> **@pvieito There are two daily checks**

Not enough accurate. You can see in the provided JSON (in case that you doubt of my previous words). Some days (10/Sept) there was more than 2 exposure checks by day. And not every 12 hours. I don't know if your statement is only valid in the latest version 1.0.7

> **@pvieito : the system would refuse the unsigned diagnosis keys, which was the bug fixed by DP3T SDK 1.2.1.**

* I guess that you refer to https://github.com/DP-3T/dp3t-sdk-ios 
* I thought that the 1.0.7 fully implements Exposure Notification Express (iOS 13.7) and from your statement I am not sure if 1.0.7 is only a recompilation using a new version of DP3T SDK
[The version 2 of the protocol is different](https://developer.apple.com/documentation/bundleresources/information_property_list/enapiversion) to provide exposure notifications to people without Radar Covid app installed. 
I hope that the 1.0.7 support all features of [Exposure Notification Express](https://developer.apple.com/documentation/exposurenotification/supporting_exposure_notifications_express) for people not interested in installing a national app in every country that visits.
  * Useful links
    * https://github.com/google/exposure-notifications-server. 
    * https://github.com/google/exposure-notifications-verification-server

* DP-3T SDK refers to startTracing and stopTracing methods.

	* The only reference to those methods in the project is RadarStatusUseCase.swift
		```
		  func changeTracingStatus(active: Bool) -> Observable<RadarStatus> {
			.create { [weak self] observer in
				if active {
					do {
						try DP3TTracing.startTracing { error in
							if let error =  error {
								self?.handle(error: error, observer: observer)
							} else {
								self?.preferencesRepository.setTracing(active: active)
								observer.onNext(.active)
								observer.onCompleted()
							}
						}

					} catch {
						self?.handle(error: error, observer: observer)
					}

				} else {
					DP3TTracing.stopTracing { error in
						if let error =  error {
							observer.onError("Error stopping tracing. : \(error)")
						} else {
							self?.preferencesRepository.setTracing(active: active)
							observer.onNext(.inactive)
							observer.onCompleted()
						}
					}

				}
				return Disposables.create()
			}

		}
		```
	* If there was an issue that forces to reinstall the app to reenable notifications this is the module to debug.

**Comment by pvieito** [comment]
> I assume that @pvieito is linked to the development team

No, I am not related in any way to the development team but I have been warning about the [DP3T SDK issue](https://github.com/RadarCOVID/radar-covid-ios/issues/4) since the repo was published which was fortunately fixed on 1.0.7.

Exposure Notification Express is a totally different system, it does not depend on apps and it is only supported by some US regions.

**Comment by pvieito** [comment]
> * [The version 2 of the protocol is different](https://developer.apple.com/documentation/bundleresources/information_property_list/enapiversion) to provide exposure notifications to people without Radar Covid app installed.

Also, this is not true as specified on the documentation, the `ENAPIVersion` 2 changes how the system calculates the user’s Exposure Risk Value and it is not related in any way to Exposure Notification Express, which is an integration completely developed server-side by Apple/Google.

Currently Radar COVID uses the `ENAPIVersion` 1 as you can see on project the [`Info.plist`](https://github.com/RadarCOVID/radar-covid-ios/blob/develop/RadarCovid/Supporting/Info.plist) as required by the [DP3T SDK](https://github.com/DP-3T/dp3t-sdk-ios#using-the-sdk).

> * I was expecting a new exposure check to report it here asap. I waited from 01:15 to 22:49 without success. I didn't wait one hour more to see if your statement "twice by day" were true. My fault.

The DP3T SDK uses the system [Background Tasks framework](https://github.com/DP-3T/dp3t-sdk-ios#background-tasks) to schedule checks [twice a day](https://github.com/DP-3T/dp3t-sdk-ios/blob/aa987057a23ec598769d500aa9d766eaa471a95c/Sources/DP3TSDK/Background/DP3TBackgroundTaskManager.swift#L159), then the system decides exactly when the app is launched.

**Comment by pvieito** [comment]
> I assume that @pvieito is linked to the development team
> 
> > **@pvieito : There is no need to reinstall the app**
> 
> The app was installed by millions of users. I don't know how many multiple cases you have seen to think that there is no need to reinstall the app. I hope you don't have the syndrome "_If it works in my iPhone and 100 more, it will work in ALL iPhones_".

About this part of the comment in particular I would like to say that we _the community_ are here trying to help and ensure Radar COVID is working properly for all so I would like for everyone to try to follow the typical [CODE OF CONDUCT](https://github.com/corona-warn-app/cwa-app-ios/blob/develop/CODE_OF_CONDUCT.md) (this one is from the German Corona-Warn-App project) and respect other people on the community.

**UPDATE**: I've just requested on [other issue](https://github.com/RadarCOVID/radar-covid-ios/issues/28) to add a Code of Conduct to the Radar COVID project.

**Comment by CarlosJimenez** [comment]
> **@pvieito : About this part of the comment in particular I would like to say that we _the community_ are here trying to help and ensure Radar COVID is working properly for all so I would like for everyone to try to follow the typical [CODE OF CONDUCT](https://github.com/corona-warn-app/cwa-app-ios/blob/develop/CODE_OF_CONDUCT.md) (this one is from the German Corona-Warn-App project) and respect other people on the community.

I'm so sorry if I have bothered you. I opened the ticket to help, not to bother anyone. Your claim that reinstallation was not necessary was contrary to my experience. I was just trying to tell you to believe me instead of denying.
Again my apologies

**Comment by borjam** [comment]
Take it easy guys we all want to help!

Now, my experience. In short, **when 1.0.6 waas released I updated and it didn't work. Some state inherited from the previous non working version rendered it unusable until I deleted the app and I reinstalled 1.0.6**.

A slightly better detailed account follows:

I had it running, I don't remember the version, until it stopped working when I updated iOS to 13.7.

Before that update to 13.7 I think I suffered another "hey, it refreshed more than two days ago!" incident, but I don't remember it well, sorry.

So, noticing it wasn't working, I deleted the app and I reinstalled it. 1.0.6 hadn't still been released.

When activating it, I noticed it took an awfully long time to activate exposure tracing. I granted the permission and it stayed there, frozen, for almost a minute before asking for permission to send notifications.

And after that, it always showed the last check on 1/7/2020. No hints of errors were shown.

But I repeat, the update didn't work well for me, for whatever reason. Some corrupted internal state prevented it from working, I don't know. I had to delete the non working version and install the updated version again.

Again, I think the worst problem is some kind of documented confirmation that it's indeed working. Especially when the application, sorry to be blunt, already has a history of silent failures.

**Comment by pvieito** [comment]
@borjam Interesting!

I just wanted to clarify my previous answer to @CarlosJimenez:

> There is no need to reinstall the app.

There is no need to reinstall the app in the context of the [DP3T SDK 1.2.1 issue](https://github.com/RadarCOVID/radar-covid-ios/issues/4) or the [Radar COVID 1.0.6 issue](https://github.com/RadarCOVID/radar-covid-ios/issues/24). Of course, there may be other currently unknown bugs, which must be addressed. Sorry for the misunderstanding.

And of course, Radar COVID should address the big issue of the silent failures, so we can quickly be aware of these type of problems without the need to be inspecting the app traffic or the system Exposure Notification check list each day.

**Comment by CarlosJimenez** [comment]
I think that I should close this ticket because the issue reported is fixed with the version 1.0.7. Now I am getting new exposure checks as before i installed iOS 13.7

In that case, I think that 2 new issues should be opened

1. **The automatic update of Radar Covid doesn't work. Even it needs a reinstallation**
    > @borjam So, noticing it wasn't working, I deleted the app and I reinstalled it. 1.0.6 hadn't still been released.

    I got the same experience with the 1.0.7 version. Apps that uses the permission of Exposure notifications maybe don't follow the normal cycle of auto updating (Permission reserved to Governments). 
    If somebody decided to test with iOS 14 he could verify if reinstallation is needed. 

2. **Radar Covid doesn't have a easy way to check if it's really working**
Until then, this is the way to detect if Radar Covid is creating exposure checks properly

    ![running](https://user-images.githubusercontent.com/1741754/93638551-c5dc2a00-f9f7-11ea-8881-a9c8a6a58a6d.jpg)

@borjam @pvieito
If you agree I will close the ticket, but I don't going to open the 2 new issues. It invest too much time to help preparing this one. But there is not any feedback from developers. Please, feel free to use the images or whatever you want or made references to this one. Once I will get your feedback I will proceed to close this issue.

**Comment by borjam** [comment]
It's fine, this issue is becoming a sort of Christmas tree.

Anyhow we don't know whether the issue I saw when updating to 1.0.6 was a fault in 1.0.6 (well, in a  way it is!) or it was a worst case triggered by the failure of the previous versions that won't be reproduced.

I don't have the means to test it properly, anyway. So, in my opinion it's fine to close it. I hope the developers have read it.
